### PR TITLE
[tune] Update API Reference Page

### PR DESCRIPTION
**Pull request by richardliaw**
# BEFORE YOU REVIEW:
Please build the documentation to review this. There are stylesheet changes and formatting shifts that are Sphinx dependent. The point of the review is to spot any visual errors - there is not that much content added in this PR.

## Why are these changes needed?
This is a first step of a much larger effort to rewrite the Tune documentation. 

This first step is just to make sure the API references are self-contained (i.e., you do not need to source dive to get Tune to work).

Later on, I'll merge information from the User Guides with this page.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: doc/source/development.rst [hunk 1]
```diff
@@ -38,6 +38,31 @@ the following:
...
 `Testing locally`_.
 
+.. _python-develop:
+
+Developing Ray (Python Only)
+----------------------------
+
+.. note:: Unless otherwise stated, directory and file paths are relative to the project root directory.
+
+RLlib, Tune, Autoscaler, and most Python files do not require you to build and compiling Ray. Follow these instructions to develop Ray's Python files locally.
```
Comment (sven1977) [inline doc/source/development.rst]: "compiling" -> "compile"

File: doc/source/development.rst [hunk 2]
```diff
@@ -38,6 +38,31 @@ the following:
...
+2. Fork and clone the project to your machine. Connect your repository to the upstream (main project) ray repository.
+
+.. code-block:: shell
+
+    git clone https://github.com/[your username]/ray.git
+    cd ray
+    git remote add upstream https://github.com/ray-project/ray.git
+    # Make sure you are up-to-date on master.
+
+4. Run ``python python/ray/setup-dev.py``. This sets up links between the ``tune`` dir (among other directories) in your local repo and the one bundled with the ``ray`` package.
```
Comment (sven1977) [inline doc/source/development.rst]: 👍

File: doc/source/development.rst [hunk 3]
```diff
@@ -38,6 +38,31 @@ the following:
...
+.. _python-develop:
+
+Developing Ray (Python Only)
+----------------------------
+
+.. note:: Unless otherwise stated, directory and file paths are relative to the project root directory.
+
+RLlib, Tune, Autoscaler, and most Python files do not require you to build and compiling Ray. Follow these instructions to develop Ray's Python files locally.
+
+1. Pip install the **latest Ray wheels.** See :ref:`install-nightlies` for instructions.
```
Comment (sven1977) [inline doc/source/development.rst]: Maybe also add a warning here not to re-do another `pip install ray` (followed by installing the wheel via `pip install -U [wheel]`). I think I did that once and it then erases lots of files (weirdly not all) in the git ray folder. One can only get out of this by rolling back the erased files with git to get back to the clean branch.

File: doc/source/tune-contrib.rst [hunk 4]
```diff
@@ -11,48 +11,12 @@ We welcome (and encourage!) all forms of contributions to Tune, including and no
...
-.. note::
-
-  If raising a new issue or PR related to Tune, be sure to include "[tune]" in the beginning of the title.
-
-For project organization, Tune maintains a relatively up-to-date organization of
-issues on the `Tune Github Project Board <https://github.com/ray-project/ray/projects/4>`__.
-Here, you can track and identify how issues are organized.
+First, following the instructions in :ref:`python-develop` to develop Tune without compiling Ray.
 
+After Ray is set up, run ``pip install -r ray/python/ray/tune/requirements-dev.txt`` to install all packages required for development of Tune.
```
Comment (sven1977) [inline doc/source/tune-contrib.rst]: "developing Tune" or "for the development of Tune"

File: doc/source/tune-contrib.rst [hunk 5]
```diff
@@ -65,11 +29,11 @@ There are a couple steps to merge a contribution.
...
      git remote add upstream https://github.com/ray-project/ray.git
      git fetch upstream
-     git rebase upstream/master
+     git rebase upstream/master # or git pull . upstream/master
 
 2. Make sure all existing tests `pass <tune-contrib.html#testing>`__.
 3. If introducing a new feature or patching a bug, be sure to add new test cases
-   in the relevant file in `tune/tests/`.
+   in the relevant file in ``tune/tests/``.
 4. Document the code. Public functions need to be documented, and remember to provide an usage
```
Comment (sven1977) [inline doc/source/tune-contrib.rst]: "a usage"

File: python/ray/tune/suggest/dragonfly.py [hunk 6]
```diff
@@ -18,7 +18,7 @@
 class DragonflySearch(SuggestionAlgorithm):
     """A wrapper around Dragonfly to provide trial suggestions.
 
-    Requires Dragonfly to be installed.
+    Requires Dragonfly to be installed via ``pip install dragonfly-opt``.
```
Comment (sven1977) [inline python/ray/tune/suggest/dragonfly.py]: 👍

File: python/ray/tune/tune.py [hunk 7]
```diff
@@ -98,12 +98,11 @@ def run(run_or_experiment,
...
         raise_on_failed_trial=True,
         return_trials=False,
-        ray_auto_init=True,
-        sync_function=None):
+        ray_auto_init=True):
     """Executes training.
 
     Args:
-        run_or_experiment (function|class|str|Experiment): If
+        run_or_experiment (function | class | str | :class:`Experiment`): If
```
Comment (sven1977) [inline python/ray/tune/tune.py]: Should we use `Union[a,b,c]` here, instead?

**Review by sven1977 — APPROVED** [review]
Nice! Just a few questions, suggestions and typo fixes.

File: doc/source/development.rst [hunk 8]
```diff
@@ -38,6 +38,31 @@ the following:
...
+.. code-block:: shell
+
+    git clone https://github.com/[your username]/ray.git
+    cd ray
+    git remote add upstream https://github.com/ray-project/ray.git
+    # Make sure you are up-to-date on master.
+
+4. Run ``python python/ray/setup-dev.py``. This sets up links between the ``tune`` dir (among other directories) in your local repo and the one bundled with the ``ray`` package.
+
+.. warning:: Do not run ``pip uninstall ray`` if setting up your environment this way. To uninstall, you must ``rm -rf`` the installation site (usually a ``site-packages/ray`` location).
```
Comment (richardliaw) [inline doc/source/development.rst]:
```suggestion
.. warning:: Do not run ``pip uninstall ray`` or ``pip install -U`` if setting up your environment this way. To uninstall or upgrade, you must ``rm -rf`` the installation site (usually a ``site-packages/ray`` location).
```

File: doc/source/development.rst (hunk 3 above)
Comment (richardliaw) [reply doc/source/development.rst]: I think this happened after you symlinked; I've added a note below as a warning.

File: python/ray/tune/tune.py (hunk 7 above)
Comment (richardliaw) [reply python/ray/tune/tune.py]: Not sure... I'm still a little unfamiliar with Unions myself. Will keep in mind.

**Pull request merged**